File: app/src/main/java/com/supercilex/robotscouter/models/scout/viewholders/SpinnerViewHolder.java
package com.supercilex.robotscouter.models.scout.viewholders;

import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import com.google.firebase.database.DatabaseReference;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.models.scout.metrics.ScoutMetric;
import com.supercilex.robotscouter.models.scout.metrics.ScoutSpinner;

public class SpinnerViewHolder extends ScoutViewHolder {
    private TextView mName;
    private Spinner mSpinner;

    public SpinnerViewHolder(View itemView) {
        super(itemView);
        mName = (TextView) itemView.findViewById(R.id.spinner_text_view);
        mSpinner = (Spinner) itemView.findViewById(R.id.spinner);
    }

    @Override
    public void initialize(ScoutMetric view, DatabaseReference ref) {
        ScoutSpinner scoutSpinner = (ScoutSpinner) view;

        setText(scoutSpinner.getName());
        initializeSpinner(scoutSpinner);
        setOnItemSelectedListener(scoutSpinner, ref);
    }

    public void setText(String name) {
        mName.setText(name);
    }

    private void initializeSpinner(ScoutSpinner scoutSpinner) {
        ArrayAdapter<String> spinnerArrayAdapter = new ArrayAdapter<>(mSpinner.getContext(),
                                                                      android.R.layout.simple_spinner_item,
                                                                      scoutSpinner.getValue());
        spinnerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        mSpinner.setAdapter(spinnerArrayAdapter);
        mSpinner.setSelection(scoutSpinner.getSelectedValue());
    }

    private void setOnItemSelectedListener(final ScoutSpinner scoutSpinner,
                                           final DatabaseReference ref) {
        mSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent,
                                       View view,
                                       int itemPosition,
                                       long id) {
                scoutSpinner.setSelectedValue(ref, itemPosition);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scouting/scoutlist/viewholder/SpinnerViewHolder.kt
package com.supercilex.robotscouter.ui.scouting.scoutlist.viewholder

import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.data.model.Metric
import com.supercilex.robotscouter.ui.scouting.MetricViewHolderBase
import kotterknife.bindView

class SpinnerViewHolder(
        itemView: View
) : MetricViewHolderBase<Metric.List, List<Metric.List.Item>, TextView>(itemView),
        AdapterView.OnItemSelectedListener {
    private val spinner: Spinner by bindView(R.id.spinner)

    init {
        spinner.adapter = ArrayAdapter<String>(
                itemView.context,
                android.R.layout.simple_spinner_item
        ).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        spinner.onItemSelectedListener = this
    }

    public override fun bind() {
        super.bind()
        updateAdapter()
        spinner.setSelection(metric.selectedValueId?.let { id ->
            metric.value.indexOfFirst { it.id == id }
        } ?: 0)
    }

    private fun updateAdapter() {
        @Suppress("UNCHECKED_CAST") // We know the metric type
        (spinner.adapter as ArrayAdapter<String>).apply {
            clear()
            addAll(metric.value.map { it.name })
        }
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, itemPosition: Int, id: Long) {
        metric.selectedValueId = metric.value[itemPosition].id
    }

    override fun onNothingSelected(view: AdapterView<*>) = Unit
}
